tests/contrib/operators/test_singularity_operator.py

```diff
@@ -0,0 +1,180 @@
+# -*- coding: utf-8 -*-
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+import unittest
+import six
+
+from parameterized import parameterized
+from spython.instance import Instance
+from airflow import AirflowException
+
+try:
+    from airflow.contrib.operators.singularity_operator import SingularityOperator
+except ImportError:
+    pass
```

Comment:
Do we need this?

Reply:
I don't think so, it's not logical that we would, if the import can't happen the test doesn't do any special checks for it.

Reply:
For the pre-commit hook, it's hanging on `Forcing response 'no' to 'rebuild CI?'` I don't think I would want to run this locally (hence answering no) but it doesn't seem to get beyond that. Is this expected?

Reply:
Let me know how you'd suggest that I proceed.

Reply:
okay I let that run and it failed, so I removed the hooks. If there needs to be some formatting let's talk about commands to run that (outside of the much more complex logic added by the pre-commit command).

Reply:
okay I let that run and it failed, so I removed the hooks. If there needs to be some formatting let's talk about commands to run that (outside of the much more complex logic added by the pre-commit command).

Reply:
The hook is often expected to fail but it will apply required changes (sort of imports, no spaces at the end of line, etc) to your file. After that you should add changed files and try to commit one more time.

It can "hang" a little bit on forcing response but then it should go on. @potiuk

Reply:
Sure I can try this again - do I select "yes" or "no" at the prompt? I believe I ultimately selected "yes" last time, it failed, and then didn't allow the commit to go through (and it didn't seem to be changed in formatting either).

Reply:
Also, do I want to install just the pre-commit or also the pre-push hooks?

Reply:
Hey @vsoch. This is not strictly required, it's just helpful for you to get changes verified without waiting for Travis build. 

It's up to you if you want to run pre-commit only for commits or also for push. The latter has the benefit that it will run checks for all commits you are pushing together in case you have more than one commit.

Here you have more infor: https://github.com/apache/airflow/blob/master/STATIC_CODE_CHECKS.rst#pre-commit-hooks

Reply:
Also regarding the pre-commit hook - answer "yes" with the first question and wait a bit to get it build (in the version that is coming soon I hope it will have more meaningful progress bar.

Reply:
Also in case of your problems in the latest build - just read the output messages in the failed "static checks". I think the messages there are pretty clear, but maybe you need some more guidance here. 

You can read more of the pre-commit checks and how we are using them in https://github.com/apache/airflow/blob/master/STATIC_CODE_CHECKS.rst#pre-commit-hooks  - there is a list of pre-commit checks there and explanation what is the expectation to .

But to additionally describe it here and give you additional  guidance - my observations for the current errors you have:

- The end-of-file/isort you can have pre-commit fixing it for you locally (and you will just have to commit it (end-of-file and isort for example). We have the requirements for isort-order of imports (isort is the tool that sorts them automatically) 

- In other cases you need to fix it because there is - sometimes potential - problem in the code (for example in pylint, or the checks for proper sorting of dependencies in setup.py. 

- In case of the sorting error - the dependencies in setup.py are alphabetically sorted so you should add your deps in sorted order.

- In case of pylint errors - you seem to have too many arguments in one of the operators - you can solve it either by decreasing the number of operators or by adding appropriate pylint disable. You can see how to do it in our docs - there is explanation when and how you should manually disable the checks and explanation how to do it: https://github.com/apache/airflow/blob/master/STATIC_CODE_CHECKS.rst#pylint-static-code-checks 

- The other "test related" pylint errors - that's a clear programming error that pylint detected - there is no "compat" in module "tests" in your import (we dropped compat in master as we dropped python 2 support) and SingularityOperator is not imported it seems.

> let's talk about commands to run that (outside of the much more complex logic added by the pre-commit command).

It is super easy: You can run those pre-commits manually (that's the power of pre-commit framework.

for example you can run:

`pre-commit run end-of-file-fixer`

for the changes you have locally added, or even 

`pre-commit run end-of-file-fixer --all-files'

And it will run the very same pre-commit checks as in Travis for all files. So you have easy way to reproduce it locally.

This is explained in detail in "Using pre-commit hooks" - including examples: https://github.com/apache/airflow/blob/master/STATIC_CODE_CHECKS.rst#using-pre-commit-hooks

Reply:
okay I'll run it again, not sure it will help...

Reply:
The command you provided didn't work, it said there weren't files. Here is the complete output of running this command again, I see the container(s) building, and then failure. Most of the checks that (I think) are supposed to give me feedback are skipped... 

https://pastebin.com/h3v7mwFy

Reply:
I'm going to manually go through the messages on Travis, and try to manually fix things. I see no other way to do this, this automated approach does not work.

Reply:
This is not logical to me - both of these files have a newline at the end, yet they are reported fixed.

![image](https://user-images.githubusercontent.com/814322/72750358-e6f32f00-3b79-11ea-867e-e0bba5fb2041.png)

Since the tests file wasn't changed, this suggests that it _doesn't_ want the newline...

Reply:
My suspicion was confirmed (no newline was actually wanted) and I was able to fix all the static issues (static check step passed!)

Reply:
> This is not logical to me - both of these files have a newline at the end, yet they are reported fixed.

It's rather logical IMHO - I think you are likely thinking about empty LINE at the end, but what the end-of-file rightfully complained about was extra EOL character - empty line and EOL character are two different things.

The end-of-file fixer will fail if it has either any number of empty lines at the end or lack of EOL character at the end of last line. Both cases are undesirable for various reasons (you cannot easily append to a file which has no EOL and extra empty line is superfluous and might hide/miss the true end of file in editor.

That's why it failed likely because you had an empty line at the end (technically empty line at the end is exactly TWO EOL characters). It expects exactly one EOL at the end of file - no more, no less.

> The command you provided didn't work
First of all - you did not run it with `--all-files` . That's why it did not fix your files - they were already committed, not staged.

As explained in our "Use pre-commits" documentation, `pre commit run` only works on staged files. In your case you tried to run in on already committed files. For that `pre-commit run --all-files` is appropriate because it runs it on all files, not only on staged files. That should have fixed your problem with automated fixing of the files. I even mentioned both of the versions in my previous comment.

For the docker build failure - you were extremely unlucky in this case. There was a release of an extra dependency that broke the master build (not our fault) - that's why you had the sql.h error. It's been fixed in one of the latest commit but you are currently behind by 26 commits. You should rebase your change to the latest master to get it working. 

You should do that anyway before merge (the best idea is to rebase often). You can see an exact way how to rebase in our documentation here: https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#how-to-rebase-pr

Reply:
Ah yes, I see the difference between newline and end of line, and also the reasons for the pre-commit run not detecting files, thanks for the clarification! And for the container build, haha. I should expect this sort of thing. Thankfully I could just use the CI output to debug and fix the issues, so the static checks and docs are all good now.

The issue remaining is my last post - the fact that we cannot test the singularity operator without installing Singularity, which is a non trivial thing to do on travis (you usually need a machine base with a lot of custom libraries and GoLang). Please let me know how to proceed.